### PR TITLE
Feature expose provider logs

### PR DESCRIPTION
**Pull request by LucasRoesler**
## Description
This creates a new logs endpoint that proxies the http request to the log provider as described in openfaas/faas#751

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

See openfaas/faas#751

## How Has This Been Tested?

This has been manually tested as well as some initial unit tests have been added to verify that it does not leak any goroutines 

This can be tested using these instructions 
- https://gist.github.com/LucasRoesler/a6d0c5e2e669825ec4ab613e3487f1e8 (installs with the updated k8s provider)
 - https://gist.github.com/LucasRoesler/ec0c365d02296c29e66be683247be8aa

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: gateway/handlers/logs.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+		}
+
+		clientConn, buf, err := hijacker.Hijack()
+		if err != nil {
+			log.Println("LogProxy: failed to hijack connection for streaming response")
+			return
+		}
+
+		defer clientConn.Close()
+		clientConn.SetWriteDeadline(time.Time{}) // allow arbitrary time between log writes
```
Comment (alexellis) [inline gateway/handlers/logs.go]: Self document that `time.Time{}` (empty time value) means there is no `WriteDeadline`.

File: gateway/handlers/logs.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			log.Printf("LogProxy: forwarding request failed: %s\n", err.Error())
+			buf.WriteString("HTTP/1.1 500 Server Error" + crlf)
+			buf.Flush()
+			return
+		}
+		defer logResp.Body.Close()
+
+		// write response headers directly to the buffer per RFC 2616
+		// https://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html
+		buf.WriteString("HTTP/1.1 " + logResp.Status + crlf)
```
Comment (alexellis) [inline gateway/handlers/logs.go]: I think there's a const for that in the `http` package, not sure if it's public, worth checking.

File: gateway/handlers/logs.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		defer cancel()
+
+		logResp, err := http.DefaultTransport.RoundTrip(logRequest)
+		if err != nil {
+			log.Printf("LogProxy: forwarding request failed: %s\n", err.Error())
+			buf.WriteString("HTTP/1.1 500 Server Error" + crlf)
+			buf.Flush()
+			return
+		}
+		defer logResp.Body.Close()
```
Comment (alexellis) [inline gateway/handlers/logs.go]: ?
```
if logResp.Body !=nil {

}
``

File: gateway/handlers/forwarding_proxy.go [hunk 4]
```diff
@@ -139,6 +139,10 @@ func copyHeaders(destination http.Header, source *http.Header) {
 		copy(vClone, v)
 		(destination)[k] = vClone
 	}
+
+	for _, h := range hopHeaders {
```
Comment (alexellis) [inline gateway/handlers/forwarding_proxy.go]: @stefanprodan PTAL

File: gateway/handlers/logs.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	"os"
+	"strings"
+	"time"
+)
+
+const crlf = "\r\n"
+const upstreamLogsEndpoint = "/system/logs"
+
+// NewLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+func NewLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
```
Comment (alexellis) [inline gateway/handlers/logs.go]: @stefanprodan This is what code looks like when WriteTimeout is enabled on the http.Server.

File: gateway/handlers/logs.go (hunk 3 above)
Comment (LucasRoesler) [reply gateway/handlers/logs.go]: I have added a comment that the nil check is not needed

File: gateway/handlers/logs.go (hunk 2 above)
Comment (LucasRoesler) [reply gateway/handlers/logs.go]: There is no public constant documented in the http package

File: gateway/handlers/logs.go (hunk 1 above)
Comment (LucasRoesler) [reply gateway/handlers/logs.go]: Doc added

**Comment by LucasRoesler** [comment]
@alexellis i have updated and validated the proxy implementation using only CloseNotifier and Flusher, I also rebased it on master to handle the merge conflict. I have not removed the write timeout yet, but I have validated that it works 100% as expected when you do remove the write timeout.

The license check is failing on the test utility I vendored to ensure the proxy endpoint does not cause any goroutine leaks.

File: gateway/handlers/forwarding_proxy.go (hunk 4 above)
Comment (alexellis) [reply gateway/handlers/forwarding_proxy.go]: Removed here -> https://github.com/openfaas/faas/pull/1146

File: gateway/handlers/logs.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
```
Comment (alexellis) [inline gateway/handlers/logs.go]: This is where we are using stdlib

File: gateway/handlers/logs.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+			return
+		}
+
+		return
+	}
+}
+
+// NewHijackLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+// This remains in the package for reference purposes, providers should instead use NewLogHandlerFunc
+func NewHijackLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
```
Comment (alexellis) [inline gateway/handlers/logs.go]: Taking over the connection.

File: gateway/handlers/logs.go [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+	"strings"
+	"time"
+)
+
+const crlf = "\r\n"
+const upstreamLogsEndpoint = "/system/logs"
+
+// NewLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+func NewLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
+	writeRequestURI := false
```
Comment (alexellis) [inline gateway/handlers/logs.go]: Initial version - bring in a time.Duration and kill the connection using `time.AfterFunc()`

File: gateway/handlers/logs.go [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+	if _, exists := os.LookupEnv("write_request_uri"); exists {
+		writeRequestURI = exists
+	}
+
+	upstreamLogProviderBase := strings.TrimSuffix(logProvider.String(), "/")
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
```
Comment (alexellis) [inline gateway/handlers/logs.go]: or `WithTimeout()`

File: gateway/handlers/logs.go (hunk 7 above)
Comment (alexellis) [inline gateway/handlers/logs.go]: Can you put this into a gist or file somewhere outside the PR?

**Comment by alexellis** [comment]
If the provider gives 404, we should return Not implemented to the CLI and handle that. i.e. the provider may not support logs message.

**Comment by LucasRoesler** [comment]
@alexellis i have updated this but haven't had a chance to fully test it because license check needs a new release

**Comment by LucasRoesler** [comment]
With my most recent commit, i have pushed a public tag of this  gateway to `theaxer/gateway:loggerhandler` and if you use the the updates on the CLI PR, you can now use these instructions to test the logs 

```
### You need 3 terminals

## Terminal 1

### You need the logs branch of the faas-cli
cd $PATH_TO_FAAS_CLI_REPO
hub pr checkout 613
make local-install

### Now go to your faas-netes repo
cd $PATH_TO_FAAS_NETES_REPO

make start-kind

cat password.txt | faas-cli login --gateway=http://localhost:31112 -u admin -s

export KUBECONFIG=$(kind get kubeconfig-path --name="${OF_DEV_ENV:-kind}")

helm upgrade openfaas --install ./chart/openfaas \
    --namespace openfaas  \
    --set basic_auth=true \
    --set openfaasImagePullPolicy=IfNotPresent \
    --set faasnetes.imagePullPolicy=IfNotPresent \
    --set gateway.image=theaxer/gateway:loggerhandler \
    --set faasnetes.image=openfaas/faas-netes:0.8.0 \
    --set functionNamespace=openfaas-fn

bash ./contrib/restart_port_forward.sh

## Terminal 2
faas-cli logs echo --gateway=http://localhost:31112

## Terminal 3
echo "`date`"  | faas-cli invoke echo --gateway=http://localhost:31112
```

File: gateway/handlers/forwarding_proxy.go (hunk 4 above)
Comment (alexellis) [reply gateway/handlers/forwarding_proxy.go]: @LucasRoesler we added this code elsewhere please can you remove this code if duplicated?

File: gateway/handlers/logs.go (hunk 8 above)
Comment (LucasRoesler) [reply gateway/handlers/logs.go]: This was implemented with timeout, as in the the next comment

**Comment by LucasRoesler** [comment]
@alexellis all of the comments should be addressed nwo

**Comment by alexellis** [comment]
Please can you run us through the decision to add `goleak` from Uber?

I.e.

* what it does?
* why we need it?
* if we need it here, where else do we need it?

**Comment by alexellis** [comment]
Please can you document `logs_provider_url` and any other configuration items in the gateway/README.md and in any other places you feel appropriate?

**Comment by alexellis** [comment]
Please can you add a unit test for readconfig for the new `LogsProviderURL` value? Use the existing tests as an example, generally it's something like: default (not set) / override and n/a.

File: gateway/Gopkg.toml [hunk 10]
```diff
@@ -27,3 +27,7 @@ ignored = ["github.com/openfaas/faas/gateway/queue"]
 [[constraint]]
   name = "github.com/openfaas/faas-provider"
   version = "0.9.1"
```
Comment (alexellis) [inline gateway/Gopkg.toml]: Would you be able to vendor 0.9.2 - the latest version? https://github.com/openfaas/faas-provider/releases/tag/0.9.2

**Comment by LucasRoesler** [comment]
@alexellis, i have updated the PR to address your comments. 

with respect to the questions about `goleak`, `goleak.VerifyNoLeaks` will count the goroutines before and after the test is run to ensure that there are no stray goroutines started by the code.  This is important for the log handler because the code starts at least one goroutine to handle the streaming the logs in the background, this is one way to help ensure that the logs handler is not leaking memory. Usage is strictly adding defer `goleak.VerifyNoLeaks(t)` to the start of the test.

**Comment by alexellis** [comment]
Given that we've adopted `goleak`, please can you document its usage in the CONTRIBUTING file? If you don't want this to be a mega-PR then raise an issue to track it now and it can be done later

**Comment by alexellis** [comment]
As per https://github.com/openfaas/faas/pull/1130#issuecomment-506805963 - let's document the choices we are making

File: gateway/README.md [hunk 11]
```diff
@@ -45,7 +45,9 @@ See the [openfaas/store](https://github.com/openfaas/store) repo for more.
 
 ## Logs
 
-Logs are available at the function level and can be accessed through Swarm or Kubernetes using native tooling. You can also install a Docker logging driver to aggregate your logs. By default functions will not write the request and response bodies to stdout. You can toggle this behaviour by setting `read_debug` for the request and `write_debug` for the response.
```
Comment (alexellis) [inline gateway/README.md]: The note on a logging aggregation tool is still valid, I'm not sure that should be deleted.

**Pull request merged**